### PR TITLE
optimize SQLParsingException stack to show the detailed parse error

### PR DESCRIPTION
**Pull request by lhyundeadsoul**
Changes proposed in this pull request:
- optimize SQLParsingException stack to show parse detail error which is essential to the developers
-
-
the result can be :
```
org.antlr.v4.runtime.misc.ParseCancellationException
	at org.antlr.v4.runtime.BailErrorStrategy.recover(BailErrorStrategy.java:51)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser.createTablespace(MySQLStatementParser.java:8445)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser.execute(MySQLStatementParser.java:1704)
	at org.apache.shardingsphere.sql.parser.mysql.parser.MySQLParser.parse(MySQLParser.java:38)
	at org.apache.shardingsphere.sql.parser.core.database.parser.SQLParserExecutor.twoPhaseParse(SQLParserExecutor.java:58)
	at org.apache.shardingsphere.sql.parser.core.database.parser.SQLParserExecutor.parse(SQLParserExecutor.java:46)
	at org.apache.shardingsphere.sql.parser.api.SQLParserEngine.parse(SQLParserEngine.java:47)
	at org.apache.shardingsphere.test.sql.parser.parameterized.engine.SQLParserParameterizedTest.parseSQLStatement(SQLParserParameterizedTest.java:78)
	at org.apache.shardingsphere.test.sql.parser.parameterized.engine.SQLParserParameterizedTest.assertSupportedSQL(SQLParserParameterizedTest.java:72)
	at jdk.internal.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: org.antlr.v4.runtime.NoViableAltException
	at org.antlr.v4.runtime.atn.ParserATNSimulator.noViableAlt(ParserATNSimulator.java:2026)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.execATN(ParserATNSimulator.java:467)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.adaptivePredict(ParserATNSimulator.java:393)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser.createTablespace(MySQLStatementParser.java:8425)
	... 36 more

org.apache.shardingsphere.sql.parser.exception.SQLParsingException: org.antlr.v4.runtime.misc.ParseCancellationException: syntax error at line 1, position 24, near [@3,24:34='ADDDATAFILE',<780>,1:24], no viable alternative at input 'CREATETABLESPACE`ts1`ADDDATAFILE'

	at org.apache.shardingsphere.sql.parser.core.database.parser.SQLParserExecutor.twoPhaseParse(SQLParserExecutor.java:67)
	at org.apache.shardingsphere.sql.parser.core.database.parser.SQLParserExecutor.parse(SQLParserExecutor.java:46)
	at org.apache.shardingsphere.sql.parser.api.SQLParserEngine.parse(SQLParserEngine.java:47)
	at org.apache.shardingsphere.test.sql.parser.parameterized.engine.SQLParserParameterizedTest.parseSQLStatement(SQLParserParameterizedTest.java:78)
	at org.apache.shardingsphere.test.sql.parser.parameterized.engine.SQLParserParameterizedTest.assertSupportedSQL(SQLParserParameterizedTest.java:72)
	at jdk.internal.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: org.antlr.v4.runtime.misc.ParseCancellationException: syntax error at line 1, position 24, near [@3,24:34='ADDDATAFILE',<780>,1:24], no viable alternative at input 'CREATETABLESPACE`ts1`ADDDATAFILE'
	at org.apache.shardingsphere.sql.parser.core.database.parser.ParserErrorListener.syntaxError(ParserErrorListener.java:48)
	at org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)
	at org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportNoViableAlternative(DefaultErrorStrategy.java:310)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:136)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser.createTablespace(MySQLStatementParser.java:8444)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser.execute(MySQLStatementParser.java:1704)
	at org.apache.shardingsphere.sql.parser.mysql.parser.MySQLParser.parse(MySQLParser.java:38)
	at org.apache.shardingsphere.sql.parser.core.database.parser.SQLParserExecutor.twoPhaseParse(SQLParserExecutor.java:65)
	... 33 more

```

### CODE REVIEW

**Comment by tuichenchuxin** [comment]
ref to #18948

**Comment by lhyundeadsoul** [comment]
> ref to #18948

hope useful to solve your issue

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java [hunk 1]
```diff
@@ -26,8 +26,12 @@
...
     
     private static final long serialVersionUID = -6408790652103666096L;
-    
+
     public SQLParsingException(final String sql) {
         super(XOpenSQLState.SYNTAX_ERROR, 11000, "You have an error in your SQL syntax: %s", sql);
     }
+
+    public SQLParsingException(final String sql, final Exception cause) {
+        super(cause, XOpenSQLState.SYNTAX_ERROR.getValue(), 11000, "You have an error in your SQL syntax: %s, Error message: %s", sql, cause.getMessage());
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java:35]: Do you think `You have an error in your SQL syntax: %s, error message: %s` is better?

Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java:35]: Please update the error code document at the same time.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java (hunk 1 above)
Comment (lhyundeadsoul) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java:35]: > Please update the error code document at the same time.

I am new to this.  Can you tell me where is the error code document?

Comment (lhyundeadsoul) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java]: > Do you think `You have an error in your SQL syntax: %s, error message: %s` is better?

Do you mean Error -> error？ 
done :）

File: shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java [hunk 2]
```diff
@@ -40,6 +40,11 @@ public ShardingSphereSQLException(final SQLState sqlState, final int vendorCode,
     }
     
     public ShardingSphereSQLException(final String sqlState, final int vendorCode, final String reason, final Object... messageArguments) {
+        this(null, sqlState, vendorCode, reason, messageArguments);
```
Comment (tuichenchuxin) [inline shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java:43]: There is no reason to set cause to null. May be not need to call super().

File: shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java [hunk 3]
```diff
@@ -40,6 +40,11 @@ public ShardingSphereSQLException(final SQLState sqlState, final int vendorCode,
...
+    }
+
+    public ShardingSphereSQLException(final Exception cause, final String sqlState, final int vendorCode, final String reason, final Object... messageArguments) {
+        super(cause);
```
Comment (tuichenchuxin) [inline shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java:47]: It's better to set Exception as the last parameter.

Comment (lhyundeadsoul) [reply shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java:47]: There can't be more pararmeter after indefinite length array

File: shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java (hunk 2 above)
Comment (lhyundeadsoul) [reply shardingsphere-infra/shardingsphere-infra-util/src/main/java/org/apache/shardingsphere/infra/util/exception/sql/ShardingSphereSQLException.java:43]: Calling Super(cause) is to transmit parent exception and display in the stacktrace

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java]: > > Please update the error code document at the same time.
> 
> I am new to this. Can you tell me where is the error code document?

@lhyundeadsoul You can refer this document——https://shardingsphere.apache.org/document/current/cn/user-manual/error-code/sql-error-code/.

Comment (lhyundeadsoul) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/exception/SQLParsingException.java]: > 
update here, Thx:
https://github.com/apache/shardingsphere/pull/20703
https://github.com/apache/shardingsphere/pull/20702

**Review by terrymanu — CHANGES_REQUESTED** [review]
One exception should only have one vender code

**Comment by terrymanu** [comment]
Any update for the issue?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/database/parser/SQLParserExecutor.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ private ParseASTNode twoPhaseParse(final String sql) {
             try {
                 return (ParseASTNode) sqlParser.parse();
             } catch (final ParseCancellationException e) {
-                throw new SQLParsingException(sql + ", " + e.getMessage());
```
Comment (lhyundeadsoul) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/database/parser/SQLParserExecutor.java:67]: We concat sql and error message in constructor parameter here, which only refers to SQL. That is not reasonable.

I think it's better to separate SQL and error message for unambiguous.

```
    public SQLParsingException(final String sql, final String detail) {
        super(XOpenSQLState.SYNTAX_ERROR, KERNEL_CODE, 0, "You have an error in your SQL syntax: %s, %s", sql, detail);
    }
```
But according to https://github.com/apache/shardingsphere/pull/20702#issuecomment-1234417673  each exception can only have one type of message. If I add the second parameter. All places that use this Exception have to be refactored. 
Is that right and necessary?

Comment (lhyundeadsoul) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/database/parser/SQLParserExecutor.java:67]: @terrymanu looking forward to your reply. Thx

**Comment by strongduanmu** [comment]
Hi @lhyundeadsoul, can you solve code conflict?

**Pull request closed without merging**